packages/trees/frontend/sirius-components-trees/src/hook/useToggleMachine.tsx

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+import { useMachine } from '@xstate/react';
+import { createMachine } from 'xstate';
+
+export const useToggleMachine = (initialActive: boolean = false): [boolean, () => void] => {
+  const [state, send] = useMachine(() =>
+    createMachine({
+      id: 'toggle',
+      initial: initialActive ? 'active' : 'inactive',
+      states: {
+        inactive: {
+          on: { TOGGLE: 'active' },
+        },
+        active: {
+          on: { TOGGLE: 'inactive' },
+        },
+      },
+    })
+  );
+
+  const isActive = state.matches('active');
+  const toggle = () => send('TOGGLE');
+
+  return [isActive, toggle];
+};
```

Comment:
First of all, such file does not need to have the `tsx` extension since it does not create JSX elements, `ts` would be just fine.

Now that I can see it in action, this feels way more complex than what we need and it not really match the expected state of the toolbar. What we need in the state of the toolbar is:

```
{
    "modalOpen": Modal | null
}
```

with `type Modal = 'NewDocument' | 'UploadDocument'`.

I don't see why we would need to create two separate state machines (which technically would allow both modals to be open at the same time). It's simple enough that a state machine is not even necessary so multiple ones is overkill. Something like this would be more than enough:

```
interface TreeToolBarState {
  modalOpen: Modal | null
}

type Modal = 'NewDocument' | 'UploadDocument'
```

And in the component:

```
const [state, setState] = useState<TreeToolBarState>({ modalOpen: null });
const openNewDocumentModal = () => setState(prevState => ({ ...prevState, modalOpen: 'NewDocument' }));
const openUploadDocumentModal = () => setState(prevState => ({ ...prevState, modalOpen: 'UploadDocument' }));
const closeModal = () => setState(prevState => ({ ...prevState, modalOpen: null }));
```